# MCPtaggR
![MCPtaggR_icon](https://github.com/tomoyukif/MCPtaggR/blob/main/inst/MCPtaggR_Icon.png?raw=true)

A package to conduct mappable-collinear-polymorphic tag genotyping.

## Introduction
MCPtaggR provides the functions to conduct mappable-collinear-polymorphic 
tag genotyping (MCPtagg). MCPtagg is a pipeline for 
next generation sequencing (NGS)-based genotyping technique to precisely 
detect SNPs and count reads on validated tags (short genome segments or 
regions) based on genome comparison information. Short NGS reads can not 
always be mapped to the proper positions on the reference genome if the reads 
were originally derived from a genome that have plenty polymorphisms and 
structural variations against the reference genome, e.g. mapping reads of a 
hybrid population that were derived from a cross between distant relatives. 
Polymorphic reads that were derived from a non-reference genome could result 
in mismapping and mapping bias. The mismapping and mappability bias
lead biased observation of mapped reads on SNP markers where 
reads of either allele are preferentially mapped. To eliminate unexpected 
detection of such error-prone markers, MCPtagg count reads only mapped on 
MCP tags that are the genomic sequences/regions on which NGS reads obtained by
RRS can be mapped uniquely and properly to call genotypes at SNPs. The list of 
MCP tags is obtained by whole genome sequence comparison between the genome 
sequences of the prarents of a given mapping/breeding population followed by 
sequencing read mappability assessment using simulated reads. 
The MCPtagg pipeline efficiently eliminates error-prone markers from your 
resultant genotype data obtained via a NGS-based genotyping technique, compared 
to the result produced by the conventional pipeline that maps reads on one
reference genome.
MCPtaggR currently supports NGS reads generated by reduced-representation 
sequence, e.g. genotyping-by-sequencing (GBS) and Restriction site associated 
DNA markers sequencing (RAD-seq), for a biparental population derived from
a cross between inbred parents. 

## Prerequisit
MCPtaggR requires some Bioconductor packages.
```
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("GenomicAlignments")
BiocManager::install("GenomicRanges")
BiocManager::install("gdsfmt")
BiocManager::install("SeqArray")
BiocManager::install("SNPRelate")
BiocManager::install("Rsubread")
BiocManager::install("Rsamtools")
BiocManager::install("DECIPHER")
BiocManager::install("BSgenome")
```

In addition, the Rsubread package requires the subread software installed.
Please check the [subread users guide](https://subread.sourceforge.net/SubreadUsersGuide.pdf).

MCPtaggR also needs MUMmer for the collinear block detection. Please visit the 
[MUMmer webpage](https://mummer4.github.io/install/install.html).

## Installation
You can install `MCPtaggR` from the GitHub repository.
```
if (!requireNamespace("devtools", quietly = TRUE))
    install.packages("devtools")
    
devtools::install_github("tomoyukif/MCPtaggR", build_vignettes = FALSE)
```

For more information see the [vignette](https://tomoyukif.github.io/MCPtaggR/).

## Citations
In preparation for publication...
