---
title: "DoSInG"
author: "Tomoyuki Furuta"
date: "Oct 28, 2021"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{DoSInG}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

\section{Introduction}
We find the regions having identical sequences locating in each genome in the same order.
In addition to find the ordered identical sequences in the genomes, we also need to list invalid genomic regions for genotyping.
The invalid genomic regions include regions which are not unique at one location in the genomes (sequences from repetitive regions) or have translocation between the genomes.
Those regions will be omitted in the genotyping step.
To build a complete list of those invalid regions, we use the results of Mummer runs with `--maxmatch` flag to find out identical sequences with any number of repetitiveness.

\section{Installation}

\section{Peform sytenic interval genotyping with SInG}
Load the `SInG` package.
```{r setup}
library(SInG)
library(rtracklayer)
```


Load output file from Mummer to R and then find syntenic intervals defined by the stretches with identical sequences located in the same order in the target genomes and also list up invalid intervals for genotyping.
```{r}
# Specify output file path as 'mummer_fn'.
mummer_fn <- "~/hdd3/kishima/A2Fs/mummer/kitaake_wk21.maxmatch.mummer"
```




```{r}
# Load and build a SInG object.
object <- mum2SInG(mummer_fn = mummer_fn)
```



Repeat elements and transposable elements identified via RepeatMasker or similar software can be also included as invalid stretches.
```{r}
# import the GFF file of repeat elements in O. sativa cv. Nipponbare identified by RepeatMasker.
filepath <- "~/hdd3/refGenome/kitaakeX/OsativaKitaake_499_v3.1.repeatmasked_assembly_v3.0.gff3.gz"
ref_rep <- readGFFAsGRanges(filepath = filepath)
# Extract repeat elements in the chromosomes not in the unanchored fragments.
ref_rep <- ref_rep[grepl("chr", tolower(as.character(seqnames(ref_rep))))]
seqlevels(ref_rep) <- seqlevelsInUse(ref_rep)
# Change and sort chromosome names
seqlevels(ref_rep) <- sub("*[^0-9]", "", sub("[^0-9]*", "", as.character(seqlevels(ref_rep))))
ref_rep <- sortSeqlevels(ref_rep)
seqlevels(ref_rep) <- seqlevels(object$ints$ref_sin)

# import the GFF file of repeat elements in O. glaberrima acc. WK21 identified by RepeatMasker.
filepath <- "~/hdd2/og_gwas/genomics/01_genomeData/final_genome/hardmask_chr/wk21_genome_chr.fa.out.gff"
alt_rep <- readGFFAsGRanges(filepath = filepath)
# Extract repeat elements in the chromosomes not in the unanchored fragments.
alt_rep <- alt_rep[grepl("chr", tolower(as.character(seqnames(alt_rep))))]
seqlevels(alt_rep) <- seqlevelsInUse(alt_rep)
# Change and sort chromosome names
seqlevels(alt_rep) <- sub("*[^0-9]", "", sub("[^0-9]*0*", "", as.character(seqlevels(alt_rep))))
alt_rep <- sortSeqlevels(alt_rep)
seqlevels(alt_rep) <- seqlevels(object$ints$alt_sin)
```

```{r}
object <- setInvalid(object, ref_rep, "ref")
object <- setInvalid(object, alt_rep, "alt")
```



Save the SInG object.
```{r}
out_fn <- "~/hdd3/kishima/A2Fs/SInG.Rdata"
save(object, file = out_fn)
```


Execute SInG for reads recorded in BAM files.
```{r}
dir <- "~/hdd3/kishima/A2Fs/aln/"
bam_files <- list.files(dir, pattern = ".bam$", full.names = TRUE)
object <- doSInG(object, bam_files)
```



Save the SInG object.
```{r}
out_fn <- "~/hdd3/kishima/A2Fs/SInG.Rdata"
save(object, file = out_fn)
```



Calculate read ratio based on normalized read counts of each interval using calcRatio(). 
The normalized read counts are the values obtained by dividing read counts by lengths of variable stretches in syntenic intervals. The variable stretches are the sequences witch differ between the given genomes and are used for distinguish which read derived from which genome. The calcRatio() function also calculate smoothed read ratio which are average read ratio within a specified window.
```{r}
object <- calcRatio(object, window = 100000, min_marker = 0)
```



Save the SInG object.
```{r}
out_fn <- "~/hdd3/kishima/A2Fs/SInG.Rdata"
save(object, file = out_fn)
```



Visualize read ratio.
```{r}
pdf(file = "~/hdd3/kishima/A2Fs/plots/A2F-1_ReadRatio.pdf")
plotReadRatio(object, ind = 1)
dev.off()

pdf(file = "~/hdd3/kishima/A2Fs/plots/A2F-2_ReadRatio.pdf")
plotReadRatio(object, ind = 2)
dev.off()

pdf(file = "~/hdd3/kishima/A2Fs/plots/A2F-3_ReadRatio.pdf")
plotReadRatio(object, ind = 3)
dev.off()

pdf(file = "~/hdd3/kishima/A2Fs/plots/A2Fs_all_ReadRatio.pdf")
plotReadRatio(object, smooth_only = TRUE)
dev.off()
```


Output genotyping result as a VCF file.
```{r}
sing2VCF(object, file_path = "~/hdd3/kishima/A2Fs/SInG_result.vcf")
sing2CSV(object, file_path = "~/hdd3/kishima/A2Fs/SInG_result.csv")
```



```{r}
sessionInfo()
```

